### ISSUE TITLE
"exists" filters are extremely slow on an object field

### ISSUE THREAD

**Issue by ssllmit** [elastic/elasticsearch#71488]
**Elasticsearch version** (`bin/elasticsearch --version`):

7.10.0

**Plugins installed**: []

**JVM version** (`java -version`):

openjdk version "1.8.0_282"

**OS version** (`uname -a` if on a Unix-like system):

Linux 3.10.0-1160.6.1.el7.x86_64

**Description of the problem including expected versus actual behavior**:

We observe that using `"exists"` with filtered queries is extremely slow.

MWE:
> `POST _search`
```json
{
  "query": {
    "bool": {
      "filter": [
        {
          "exists": {
            "field": "my_object_field_name"
          }
        }
      ]
    }
  },
  "size": 1,
  "track_total_hits": true
}
```

In contrast, keyword matches are extremely fast. Example:
```json
{
  "query": {
    "bool": {
      "filter": [
        {
          "match": {
            "my_object_field_name.prop1": "my-keyword-value"
          }
        }
      ]
    }
  },
  "size": 1,
  "track_total_hits": true
}
```

I've read these issues, and believe that this is not caused by using a `"must"`, which computes a score for every document.
* #22657
* https://discuss.elastic.co/t/very-slow-performance-of-must-not-with-exists-in-filter/144281

Indeed, even assigning random scores to get a random filtered document is very fast:
```json
{
  "query": {
    "bool": {
      "filter": [
        {
          "match": {
            "my_object_field_name.prop1": "my-keyword-value"
          }
        }
      ],
      "must": [
        {
          "function_score": {
            "functions": [
              {
                "random_score": {}
              }
            ],
            "boost_mode": "replace"
          }
        }
      ]
    }
  },
  "size": 1,
  "track_total_hits": true
}
```

Adding the `"exists"` condition above to any of these queries results in extraordinarily slow performance and sometimes even node timeouts.

Perhaps an existence query like this should be expressed as a `"should"` with all possible subfield names. This example is fast, but seems at odds with the pointers above about not matching and scoring all documents with a `"should"`:
```json
{
  "query": {
    "bool": {
      "should": [
        {
          "exists": {
            "field": "my_object_field_name.prop1"
          }
        },
        {
          "exists": {
            "field": "my_object_field_name.prop2"
          }
        }
      ],
      "minimum_should_match": 1,
      "must": [
        {
          "function_score": {
            "functions": [
              {
                "random_score": {}
              }
            ],
            "boost_mode": "replace"
          }
        }
      ]
    }
  },
  "size": 1,
  "track_total_hits": true
}
```

This behavior with `"exists"` on object fields is unexpected and took a while to track down.

**Comment by nik9000** [comment]
I *thought* we had another issue for `exists` on `object` fields being slow. It makes a big `this-exists OR that-exists OR the-other-exists` query. I'll see if I can dig it up.

**Comment by nik9000** [comment]
> Perhaps an existence query like this should be expressed as a `"should"` with all possible subfield names.

I believe that is how we implement `exists` for `object` fields [already](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/query/ExistsQueryBuilder.java#L169). So:
```
{
  "query": {
    "bool": {
      "filter": [
        {
          "exists": {
            "field": "my_object_field_name"
          }
        }
      ]
    }
  }
}
```

Should be executed as though:
```
{
  "query": {
    "bool": {
      "filter": [
        {
          "bool": {
            "should": [
              {
                "exists": {
                  "field": "my_object_field_name.prop1"
                }
              },
              {
                "exists": {
                  "field": "my_object_field_name.prop2"
                }
              }
            ],
            "minimum_should_match": 1
          }
        ]
      }
    }
}
```

I think #64837 is the issue I was thinking of. Do you think it's reasonable to move this conversation there?

**Comment by nik9000** [comment]
I'll close this issue now and consider this merged into #64837.
